plugins/search-backend-module-elasticsearch/src/engines/ElasticSearchSearchEngine.ts

```diff
@@ -0,0 +1,265 @@
+      });
+    }
+    logger.info('Initializing ElasticSearch search engine.');
+    return new Client({
+      node: config.getString('node'),
+      auth: config.get<ElasticConfigAuth>('auth'),
+    });
+  }
+
+  protected translator({
+    term,
+    filters = {},
+    types,
+  }: SearchQuery): ConcreteElasticSearchQuery {
+    return {
+      elasticQueryBuilder: () => {
+        const filter = Object.entries(filters)
+          .filter(([_, value]) => Boolean(value))
+          .map(([key, value]: [key: string, value: any]) => {
+            if (['string', 'number', 'boolean'].includes(typeof value)) {
+              return esb.matchQuery(key, value.toString());
+            }
+            if (Array.isArray(value)) {
+              return esb
+                .boolQuery()
+                .should(value.map(it => esb.matchQuery(key, it.toString())));
+            }
+            this.logger.error(
+              'Failed to query, unrecognized filter type',
+              key,
+              value,
+            );
+            throw new Error(
+              'Failed to add filters to query. Unrecognized filter type',
+            );
+          });
+        const query = isBlank(term)
+          ? esb.matchAllQuery()
+          : esb
+              .multiMatchQuery(['*'], term)
+              .fuzziness('auto')
+              .minimumShouldMatch(1);
+        return (
+          esb
+            .requestBodySearch()
+            .query(esb.boolQuery().filter(filter).must([query]))
+            // TODO: Replace size limit with page cursor after pagination approach decided
+            // See: https://github.com/backstage/backstage/issues/6062
+            .size(100)
+        );
+      },
+      documentTypes: types,
+    };
+  }
+
+  setTranslator(translator: ElasticSearchQueryTranslator) {
+    this.translator = translator;
+  }
+
+  async index(type: string, documents: IndexableDocument[]): Promise<void> {
+    this.logger.info(
+      `Started indexing ${documents.length} documents for index ${type}`,
+    );
+    const startTimestamp = process.hrtime();
+    const alias = this.constructSearchAlias(type);
+    try {
+      const aliases = await this.elasticSearchClient.cat.aliases({
+        format: 'json',
+        name: alias,
+      });
+      const removableIndices = aliases.body.map(
+        (r: Record<string, any>) => r.index,
+      );
+      const index = this.constructIndexName(type, `${Date.now()}`);
+
+      await this.elasticSearchClient.indices.create({
+        index,
+      });
+      const result = await this.elasticSearchClient.helpers.bulk({
+        datasource: documents,
+        onDocument() {
+          return {
+            index: { _index: index },
+          };
+        },
+        refreshOnCompletion: index,
+      });
+
+      this.logger.info(
+        `Indexing completed for index ${type} in ${duration(startTimestamp)}`,
+        result,
+      );
+      await this.elasticSearchClient.indices.updateAliases({
```

Comment:
Is it possible that this operation will fail if the index to be removed does not exist? I mean, what happens if you run the indexer for the first time on a fresh instance of Elastic? Should we use `try/catch` to roll things back / clean up or maybe use `indices.existsAlias` somewhere here?

Reply:
This particular case couldn't actually happen with the current flow of actions within this indexing process. We are creating the index three lines above this one so at least one item of this wildcarded index would always be present, unless the index creation fails which would be a bigger issue. 

The whole indexing is wrapped into a try/catch but I'm not certain what clean up jobs we could do in case of error. Note that the last two actions (our of total four actions) are destructive and rollback from those would be tricky. 
The remaining two actions are creating the index and bulk inserting data into it. If index creation fails, we'll do nothing regardless. If bulk insertion fails (completely, partial success is considered success) the index is left behind as empty and removed during the next indexing run. 

Might be worth removing that though in case the failure case continues during those scheduled runs, so I'll add an index deletion to be part of the catch clause, in case newly created one exists.